inst/odin/compare_carehomes.cpp

```diff
@@ -0,0 +1,215 @@
+  if (std::isnan(data_x) || std::isnan(data_size)) {
+    return 0;
+  }
+  return dust::dbetabinom(data_x, data_size, model_prob, rho, true);
+}
+
+template <typename real_t>
+real_t test_prob_pos(real_t pos, real_t neg, real_t sensitivity,
+                     real_t specificity, real_t exp_noise,
+                     dust::rng_state_t<real_t>& rng_state) {
+  // We add some exponential noise to the number of positives and negatives
+  // to help ensure prob_pos is not 0 or 1. If e.g. prob_pos were 0 and there
+  // were individuals who tested positive, this would result in a weight of 0
+  // for a particle. If all particles have weights of 0, the particle filter
+  // breaks. The exponential noise produces small non-zero weights in these
+  // circumstances to prevent the particle filter from breaking.
+  pos += dust::distr::rexp(rng_state, exp_noise);
+  neg += dust::distr::rexp(rng_state, exp_noise);
+  return (sensitivity * pos + (1 - specificity) * neg) / (pos + neg);
+}
+
+// Lots of data items here. Even though some of these are counts (in
+// fact, most of them are counts) we'll pull them in as doubles
+// because that will help with the NA handling. Once this is working
+// reliably we could cast these explicitly as ints, but there is not
+// likely to be any major gain here.
+//
+// [[odin.dust::compare_data(icu = double)]]
+// [[odin.dust::compare_data(general = double)]]
+// [[odin.dust::compare_data(hosp = double)]]
+// [[odin.dust::compare_data(deaths_hosp = double)]]
+// [[odin.dust::compare_data(deaths_comm = double)]]
+// [[odin.dust::compare_data(deaths = double)]]
+// [[odin.dust::compare_data(admitted = double)]]
+// [[odin.dust::compare_data(new_conf = double)]]
+// [[odin.dust::compare_data(new_admitted = double)]]
+// [[odin.dust::compare_data(npos_15_64 = double)]]
+// [[odin.dust::compare_data(ntot_15_64 = double)]]
+// [[odin.dust::compare_data(pillar2_pos = double)]]
+// [[odin.dust::compare_data(pillar2_tot = double)]]
+// [[odin.dust::compare_data(pillar2_cases = double)]]
+// [[odin.dust::compare_data(pillar2_over25_pos = double)]]
+// [[odin.dust::compare_data(pillar2_over25_tot = double)]]
+// [[odin.dust::compare_data(pillar2_over25_cases = double)]]
+// [[odin.dust::compare_data(react_pos = double)]]
+// [[odin.dust::compare_data(react_tot = double)]]
+// [[odin.dust::compare_data(strain_non_variant = double)]]
+// [[odin.dust::compare_data(strain_tot = double)]]
+// [[odin.dust::compare_function]]
+template <typename T>
+typename T::real_t compare(const typename T::real_t * state,
+                           const typename T::data_t& data,
+                           const typename T::internal_t internal,
+                           std::shared_ptr<const typename T::shared_t> shared,
+                           dust::rng_state_t<typename T::real_t>& rng_state) {
+  typedef typename T::real_t real_t;
+
+  // State variables; these largely correspond to the quantities in data
```

Comment:
Why are some unpacked here, and some unpacked where they are used?

Reply:
That just mimics how the original compare function was written to make the connection with the R easier https://github.com/mrc-ide/sircovid/blob/i200-compiled-compare/R/carehomes.R#L514-L527